With 32-bit python on 64-bit Windows, we get the following behavior.

(NOTE: I added a `ctx.msg()` that displays MSVC targets, for easier debugging)

---
Default:

```
python.exe waf configure
Setting top to                           : C:\Users\Rok\Development\pyi-bootloader\pyinstaller\bootloader
Setting out to                           : C:\Users\Rok\Development\pyi-bootloader\pyinstaller\bootloader\build
Python Version                           : 3.12.2 (tags/v3.12.2:6abddd9, Feb  6 2024, 21:15:12) [MSC v.1937 32 bit (Intel)]
MSVC target(s)                           : ['x86', 'arm64_x86', 'amd64_x86', 'x86_x86']
Checking for 'msvc' (C compiler)         : C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.39.33519\bin\HostX86\x86\CL.exe
Platform                                 : Windows-32bit-intel detected based on compiler
Build tests                              : disabled
Checking for function unsetenv           : no
Checking for function mkdtemp            : no
Checking for function dirname            : no
Checking for function basename           : no
Checking for function strnlen            : yes
'configure' finished successfully (4.047s)
```

This looks good (aside from one invalid MSVC target, but I'll comment on that separately). We are using the native x86 toolchain, and we have detected 32-bit Intel. So this would build.

---
Trying to force 64-bit build via `--target-arch`:

```
python.exe waf configure --target-arch 64bit
Setting top to                           : C:\Users\Rok\Development\pyi-bootloader\pyinstaller\bootloader
Setting out to                           : C:\Users\Rok\Development\pyi-bootloader\pyinstaller\bootloader\build
Python Version                           : 3.12.2 (tags/v3.12.2:6abddd9, Feb  6 2024, 21:15:12) [MSC v.1937 32 bit (Intel)]
MSVC target(s)                           : ['x64', 'arm64_x64', 'amd64_x64', 'x86_x64']
Checking for 'msvc' (C compiler)         : C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.39.33519\bin\HostX64\x64\CL.exe
Platform                                 : Windows-64bit manually chosen
Build tests                              : disabled
Checking for function unsetenv           : no
Checking for function mkdtemp            : no
Checking for function dirname            : no
Checking for function basename           : no
Checking for function strnlen            : yes
'configure' finished successfully (4.373s)
```

Also good - native x64 toolchain, and target is identified as 64-bit Intel.

---
Trying to force 64-bit ARM build via `--target-arch`:

```
python.exe waf configure --target-arch 64bit-arm
Setting top to                           : C:\Users\Rok\Development\pyi-bootloader\pyinstaller\bootloader
Setting out to                           : C:\Users\Rok\Development\pyi-bootloader\pyinstaller\bootloader\build
Python Version                           : 3.12.2 (tags/v3.12.2:6abddd9, Feb  6 2024, 21:15:12) [MSC v.1937 32 bit (Intel)]
MSVC target(s)                           : ['arm64', 'arm64_arm64', 'amd64_arm64', 'x86_arm64']
Checking for 'msvc' (C compiler)         : C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.39.33519\bin\HostX64\ARM64\CL.exe
Platform                                 : Windows-64bit-arm manually chosen
Build tests                              : disabled
Checking for function unsetenv           : no
Checking for function mkdtemp            : no
Checking for function dirname            : no
Checking for function basename           : no
Checking for function strnlen            : yes
'configure' finished successfully (8.227s)
```

Good - toolchain for x64 host and ARM64 target; target is identified as 64-bit ARM.

Unfortunately, before I installed amd64_arm64 toolchain, we ended up picking the wrong compiler:

```
python.exe waf configure --target-arch 64bit-arm
Setting top to                           : C:\Users\Rok\Development\pyi-bootloader\pyinstaller\bootloader
Setting out to                           : C:\Users\Rok\Development\pyi-bootloader\pyinstaller\bootloader\build
Python Version                           : 3.12.2 (tags/v3.12.2:6abddd9, Feb  6 2024, 21:15:12) [MSC v.1937 32 bit (Intel)]
MSVC target(s)                           : ['arm64', 'arm64_arm64', 'amd64_arm64', 'x86_arm64']
Checking for 'msvc' (C compiler)         : C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.39.33519\bin\HostX64\x64\CL.exe
Platform                                 : Windows-64bit-arm manually chosen
Build tests                              : disabled
Checking for function unsetenv           : no
Checking for function mkdtemp            : no
Checking for function dirname            : no
Checking for function basename           : no
Checking for function strnlen            : no
'configure' finished successfully (7.099s)
```

In this case, the native x64 compiler was picked up, and because we believed we were building form 64-bit ARM, the build fails with
```
src\main.c.3.o : fatal error LNK1112: module machine type 'x64' conflicts with target machine type 'ARM64'
```

I need to take a closer look at what exactly went wrong with compiler detection and what we can do about it.

---
What does not work correctly yet, is inferring target from `--msvc_targets`. For example:

```
python.exe waf configure --msvc_targets x64
Setting top to                           : C:\Users\Rok\Development\pyi-bootloader\pyinstaller\bootloader
Setting out to                           : C:\Users\Rok\Development\pyi-bootloader\pyinstaller\bootloader\build
Python Version                           : 3.12.2 (tags/v3.12.2:6abddd9, Feb  6 2024, 21:15:12) [MSC v.1937 32 bit (Intel)]
MSVC target(s)                           : ['x64']
Checking for 'msvc' (C compiler)         : C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.39.33519\bin\HostX64\x64\CL.exe
Platform                                 : Windows-32bit-intel detected based on compiler
Build tests                              : disabled
Checking for function unsetenv           : no
Checking for function mkdtemp            : no
Checking for function dirname            : no
Checking for function basename           : no
Checking for function strnlen            : no
'configure' finished successfully (4.284s)
```

```
python.exe waf configure --msvc_targets amd64_arm64
Setting top to                           : C:\Users\Rok\Development\pyi-bootloader\pyinstaller\bootloader
Setting out to                           : C:\Users\Rok\Development\pyi-bootloader\pyinstaller\bootloader\build
Python Version                           : 3.12.2 (tags/v3.12.2:6abddd9, Feb  6 2024, 21:15:12) [MSC v.1937 32 bit (Intel)]
MSVC target(s)                           : ['amd64_arm64']
Checking for 'msvc' (C compiler)         : C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.39.33519\bin\HostX64\ARM64\CL.exe
Platform                                 : Windows-32bit-intel detected based on compiler
Build tests                              : disabled
Checking for function unsetenv           : no
Checking for function mkdtemp            : no
Checking for function dirname            : no
Checking for function basename           : no
Checking for function strnlen            : no
'configure' finished successfully (3.672s)
```

In both cases, we picked up the specified compiler toolchain (native x64, or amd64_arm64), but we mis-identified the target as 32-bit Intel.

-----

With 64-bit python on 64-bit Windows, we get similar behavior:

---
Default

```
python.exe waf configure
Setting top to                           : C:\Users\Rok\Development\pyi-bootloader\pyinstaller\bootloader
Setting out to                           : C:\Users\Rok\Development\pyi-bootloader\pyinstaller\bootloader\build
Python Version                           : 3.12.1 (tags/v3.12.1:2305ca5, Dec  7 2023, 22:03:25) [MSC v.1937 64 bit (AMD64)]
MSVC target(s)                           : ['x64', 'arm64_x64', 'amd64_x64', 'x86_x64']
Checking for 'msvc' (C compiler)         : C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.39.33519\bin\HostX64\x64\CL.exe
Platform                                 : Windows-64bit-intel detected based on compiler
Build tests                              : disabled
Checking for function unsetenv           : no
Checking for function mkdtemp            : no
Checking for function dirname            : no
Checking for function basename           : no
Checking for function strnlen            : yes
'configure' finished successfully (6.987s)
```

Good; native x64 toolchain, target identified as 64-bit Intel.

---

Trying to force 32-bit build via `--target-arch`:

```
C:\Users\Rok\Development\pyi-bootloader\pyinstaller\bootloader>c:\Users\Rok\AppData\Local\Programs\Python\Python312\python.exe waf configure --target-arch 32bit
Setting top to                           : C:\Users\Rok\Development\pyi-bootloader\pyinstaller\bootloader
Setting out to                           : C:\Users\Rok\Development\pyi-bootloader\pyinstaller\bootloader\build
Python Version                           : 3.12.1 (tags/v3.12.1:2305ca5, Dec  7 2023, 22:03:25) [MSC v.1937 64 bit (AMD64)]
MSVC target(s)                           : ['x86', 'arm64_x86', 'amd64_x86', 'x86_x86']
Checking for 'msvc' (C compiler)         : C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.39.33519\bin\HostX86\x86\CL.exe
Platform                                 : Windows-32bit manually chosen
Build tests                              : disabled
Checking for function unsetenv           : no
Checking for function mkdtemp            : no
Checking for function dirname            : no
Checking for function basename           : no
Checking for function strnlen            : yes
'configure' finished successfully (5.249s)
```

Good; native x86 toolchain, target identified as 32-bit Intel.

---

Trying to force 64-bit ARM build via `--target-arch`:


```
python.exe waf configure --target-arch 64bit-arm
Setting top to                           : C:\Users\Rok\Development\pyi-bootloader\pyinstaller\bootloader
Setting out to                           : C:\Users\Rok\Development\pyi-bootloader\pyinstaller\bootloader\build
Python Version                           : 3.12.1 (tags/v3.12.1:2305ca5, Dec  7 2023, 22:03:25) [MSC v.1937 64 bit (AMD64)]
MSVC target(s)                           : ['arm64', 'arm64_arm64', 'amd64_arm64', 'x86_arm64']
Checking for 'msvc' (C compiler)         : C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.39.33519\bin\HostX64\ARM64\CL.exe
Platform                                 : Windows-64bit-arm manually chosen
Build tests                              : disabled
Checking for function unsetenv           : no
Checking for function mkdtemp            : no
Checking for function dirname            : no
Checking for function basename           : no
Checking for function strnlen            : yes
'configure' finished successfully (9.168s)
```

Good - toolchain for x64 host and ARM64 target; target is identified as 64-bit ARM. (Although it probably also mis-selects compiler if amd64_arm64 toolchain is not installed).

---

And with `--msvc_targets`:

```
python.exe waf configure --msvc_targets x86
Setting top to                           : C:\Users\Rok\Development\pyi-bootloader\pyinstaller\bootloader
Setting out to                           : C:\Users\Rok\Development\pyi-bootloader\pyinstaller\bootloader\build
Python Version                           : 3.12.1 (tags/v3.12.1:2305ca5, Dec  7 2023, 22:03:25) [MSC v.1937 64 bit (AMD64)]
MSVC target(s)                           : ['x86']
Checking for 'msvc' (C compiler)         : C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.39.33519\bin\HostX86\x86\CL.exe
Platform                                 : Windows-64bit-intel detected based on compiler
Build tests                              : disabled
Checking for function unsetenv           : no
Checking for function mkdtemp            : no
Checking for function dirname            : no
Checking for function basename           : no
Checking for function strnlen            : no
'configure' finished successfully (4.353s)
```

```
python.exe waf configure --msvc_targets amd64_arm64
Setting top to                           : C:\Users\Rok\Development\pyi-bootloader\pyinstaller\bootloader
Setting out to                           : C:\Users\Rok\Development\pyi-bootloader\pyinstaller\bootloader\build
Python Version                           : 3.12.1 (tags/v3.12.1:2305ca5, Dec  7 2023, 22:03:25) [MSC v.1937 64 bit (AMD64)]
MSVC target(s)                           : ['amd64_arm64']
Checking for 'msvc' (C compiler)         : C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.39.33519\bin\HostX64\ARM64\CL.exe
Platform                                 : Windows-64bit-intel detected based on compiler
Build tests                              : disabled
Checking for function unsetenv           : no
Checking for function mkdtemp            : no
Checking for function dirname            : no
Checking for function basename           : no
Checking for function strnlen            : no
'configure' finished successfully (4.168s)
```

So again, we pick up correct toolchain (native x86, or amd64_arm64), but target is mis-identified as 64-bit Intel.

----

I think the changes made to `detect_arch` are actually hurting us. We should be inferring those settings from compiler after all, after ensuring that compiler is correctly selected:

1. at start of `configure`, we should set `ctx.env['_MSVC_TARGETS']`, as we already do:
https://github.com/pyinstaller/pyinstaller/blob/18d5325e871865e24f194a04ebc88968ad84a0ee/bootloader/wscript#L406

I think it would also help if we displayed the MSVC target(s) here, to make things easier to debug. Maybe put it under a check - if we're on Windows and we are not preferring gcc or clang, display MSVC targets with `ctx.message`. At this point, `ctx.env.DEST_OS` is not available yet, so we will probably have to check the platform that python is running on.

2. Our `_msvc_target` should correctly narrow the compiler selection - which current changeset already does. I.e., honor `--msvc_targets` if given, otherwise honor `--target-arch` if given, else fall back to python's architecture.

Long-term, we should probably ensure that we do not generate invalid names for _MSVC_TARGETS, such as `arm64_arm64` and `x86_x86`, `amd64_x64`.  Or even `x86_x64` (which should be `x86_amd64`).

3. In `detect_arch` (which is called in `configure` after we attempt to load/select compiler), we should infer settings from the compiler, instead from python (the changes made there should be reverted).

4. In `set_arch_flags` that follows, we should remove this call to `_msvc_target`
https://github.com/pyinstaller/pyinstaller/blob/18d5325e871865e24f194a04ebc88968ad84a0ee/bootloader/wscript#L288

as it is useless at this point - the `_MSVC_TARGETS` has already been set at the start of `configure`, and it also has already been used to select compiler.

--- a/PyInstaller/compat.py
+++ b/PyInstaller/compat.py
@@ -231,7 +231,12 @@ else:
 system = 'Cygwin' if is_cygwin else platform.system()
 
 # Machine suffix for bootloader.
-machine = _pyi_machine(platform.machine(), platform.system())
+if is_win:
+    # On Windows ARM64 using an x64 Python environment, platform.machine() returns ARM64 but
+    # we really want the bootloader that matches the Python environment instead of the OS.
+    machine = _pyi_machine(os.environ.get("PROCESSOR_ARCHITECTURE", platform.machine()), platform.system())
+else:
+    machine = _pyi_machine(platform.machine(), platform.system())
 
 
 # Wine detection and support

--- a/bootloader/wscript
+++ b/bootloader/wscript
@@ -379,31 +379,52 @@ int main ()
 
 
 def _msvc_target(ctx):
+    if getattr(ctx.options, "msvc_targets", False):
+        return re.findall(r"[^, ]+", ctx.options.msvc_targets)
+
     try:
-        host = {"AMD64": "amd64", "x86": "x86", "ARM64": "arm64"}[platform.machine()]
+        _msvc_names = {"AMD64": "amd64", "x86": "x86", "ARM64": "arm64"}
+        host = _msvc_names[os.environ.get("PROCESSOR_ARCHITECTURE", platform.machine())]
     except KeyError:
         return []
+
     if ctx.options.target_arch == '32bit':
         target = 'x86'
     elif ctx.options.target_arch == '64bit':
-        target = 'amd64'
+        target = 'x64'
     elif ctx.options.target_arch == '64bit-arm':
         target = 'arm64'
     else:
-        return []
-    # If we are not cross-compiling, we need to return single-architecture target string.
-    if host == target:
-        if host == 'amd64':
-            return 'x64'  # Native amd64 compiler is actually called 'x64'. See `waflib/Tools/msvc.py`.
-        return host  # x86, arm64
-    # Cross-compiling; set host_target combination: amd64_x86, amd64_arm64, x86_amd64, arm64_amd64, ...
-    return [host + "_" + target]
+        target = "x64" if host == "amd64" else host
+
+    if host == 'x86':
+        if target == 'x86':
+            return ['x86']
+        elif target == 'x64':
+            return ['x86_amd64']
+        else:  # arm64
+            return ['x86_arm64']
+    elif host == 'amd64':
+        if target == 'x86':
+            return ['amd64_x86', 'x86']
+        elif target == 'x64':
+            return ['x64', 'x86_amd64']
+        else:  # arm64
+            return ['amd64_arm64', 'x86_arm64']
+    else:  # arm64
+        if target == 'x86':
+            return ['arm64_x86', 'x86', 'amd64_x86']
+        elif target == 'x64':
+            return ['arm64_amd64', 'x64', 'x86_amd64']
+        else:  # arm64
+            return ['arm64']
 
 
 def configure(ctx):
     ctx.msg('Python Version', sys.version.replace(os.linesep, ''))
     # For MSVC the target arch must have already been set when the compiler is searched.
     ctx.env['MSVC_TARGETS'] = _msvc_target(ctx)
+    ctx.msg('MSVC target(s)', ctx.env['MSVC_TARGETS'])
 
     # ** C compiler **
 

--- a/news/8219.bugfix.rst
+++ b/news/8219.bugfix.rst
@@ -1,0 +1,1 @@
+Fix bootloaders not being found when running an Intel build of Python on Windows ARM64.

--- a/news/8219.build.rst
+++ b/news/8219.build.rst
@@ -1,0 +1,2 @@
+The target architecture on Windows using MSVC now defaults to that of the
+current Python environment – not the current OS.

